substrate/frame/contracts/uapi/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+}
+
+impl ReturnFlags {
+	/// Initialize [`ReturnFlags`] with the reverted flag.
+	pub fn new_with_reverted(has_reverted: bool) -> Self {
+		Self::default().set_reverted(has_reverted)
+	}
+
+	/// Sets the bit to indicate that the execution is going to be reverted.
+	#[must_use]
+	pub fn set_reverted(mut self, has_reverted: bool) -> Self {
+		match has_reverted {
+			true => self.value |= has_reverted as u32,
+			false => self.value &= !has_reverted as u32,
+		}
+		self
+	}
+
+	/// Returns the underlying `u32` representation.
+	#[cfg(not(feature = "std"))]
+	pub(crate) fn into_u32(self) -> u32 {
+		self.value
+	}
+}
+
+/// Thin-wrapper around a `u32` representing a pointer for Wasm32.
+///
+/// Only for shared references.
+///
+/// # Note
+///
+/// Can only be constructed from shared reference types and encapsulates the
+/// conversion from reference to raw `u32`.
+/// Does not allow accessing the internal `u32` value.
+#[derive(Debug, Encode)]
+#[repr(transparent)]
+pub struct Ptr32<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a T>,
+}
+
+impl<'a, T> Ptr32<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given shared slice.
+	pub fn from_slice(slice: &'a [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+/// Thin-wrapper around a `u32` representing a pointer for Wasm32.
+///
+/// Only for exclusive references.
+///
+/// # Note
+///
+/// Can only be constructed from exclusive reference types and encapsulates the
+/// conversion from reference to raw `u32`.
+/// Does not allow accessing the internal `u32` value.
+#[derive(Debug, Encode)]
+#[repr(transparent)]
+pub struct Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a mut T>,
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given exclusive slice.
+	pub fn from_slice(slice: &'a mut [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: Sized,
+{
+	/// Creates a new Wasm32 pointer from the given exclusive reference.
+	pub fn from_ref(a_ref: &'a mut T) -> Self {
+		let a_ptr: *mut T = a_ref;
+		Self::new(a_ptr as u32)
+	}
+}
```

Comment:
Let's get rid of those silly wrappers. They are just noise to me. Let's just express our intention properly: With pointers. AFAIK they are only used in the wasm32 module anyways.